hail/python/hail/vds/impex.py

```diff
@@ -0,0 +1,335 @@
+       the struct onto the entries::
+
+            flat_gvcf_info = {f'gi_{name}': vds.variant_data.gvcf_info[name] for name in vds.variant_data.gvcf_info}
+            vds.variant_data = vds.variant_data.transmute_entries(**flat_gvcf_info)
+
+    Parameters
+    ----------
+    dataset : :class:`.MatrixTable`
+        Dataset.
+    output : :class:`str`
+        Path of .vcf or .vcf.bgz file to write.
+    append_to_header : :class:`str`, optional
+        Path of file to append to VCF header.
+    parallel : :class:`str`, optional
+        If ``'header_per_shard'``, return a set of VCF files (one per
+        partition) rather than serially concatenating these files. If
+        ``'separate_header'``, return a separate VCF header file and a set of
+        VCF files (one per partition) without the header. If ``None``,
+        concatenate the header and all partitions into one VCF file.
+    metadata : :obj:`dict` [:obj:`str`, :obj:`dict` [:obj:`str`, :obj:`dict` [:obj:`str`, :obj:`str`]]], optional
+        Dictionary with information to fill in the VCF header. See
+        :func:`get_vcf_metadata` for how this
+        dictionary should be structured.
+    tabix : :obj:`bool`, optional
+        If true, writes a tabix index for the output VCF.
+        **Note**: This feature is experimental, and the interface and defaults
+        may change in future versions.
+    """
+    ref, var = dataset.reference_data, dataset.variant_data
+
+    # fix up ref data
+    if 'END' in ref.entry:
+        ref = ref.drop('END')
+
+    if 'gvcf_info' in var.entry and isinstance(var.gvcf_info.dtype, tstruct):
+        warning(
+            'vds.export_vcf: dropping `gvcf_info` as exporting entry field of type `struct<...>` '
+            'to VCF is not supported. To preserve this information, flatten the struct onto the '
+            'entries.'
+        )
+        var = var.drop('gvcf_info')
+
+    if 'LGT' in var.entry:
+        if 'GT' not in var.entry:
+            var = var.annotate_entries(GT=lgt_to_gt(var.LGT, var.LA))
+        var = var.drop('LGT')
+    if 'LPGT' in var.entry:
+        if 'PGT' not in var.entry:
+            var = var.annotate_entries(PGT=lgt_to_gt(var.LPGT, var.LA))
+        var = var.drop('LPGT')
+
+    if 'LA' in var.entry:
+        var = var.transmute_entries(LAA=var.LA[1:])
+
+    # TODO(chrisvittal):
+    #   In a future enhancement, add appropriate metadata for all 'local' fields. We do not do this
+    #   at this time (Oct. 2024), as htsjdk, which we use for VCF header parsing, does not support
+    #   the vcf 4.5 changes necessary for completely correct metadata.
+
+    extra_header = (
+        '##SVCR="This is an VCF that implements hail\'s Scalable Variant Call '
+        'Representation. See https://doi.org/10.1101/2024.01.09.574205 for more '
+        'information."\n'
+    )
+    if VariantDataset.ref_block_max_length_field in ref.globals:
+        rbml = hl.eval(ref[VariantDataset.ref_block_max_length_field])
```

Comment:
I don't think it's necessary. We do optimize away the entire computation if we can simply read one global field. We're not _that_ bad.

Reply:
I'm not certain that's always guaranteed (what if we add a global field as the result of an aggregation?), but I think you're right that it would be rare enough we should scare people into writing when it isn't necessary, especially in this case when we can be pretty certain the max length field comes from disk.